internal/dataplane/kong_client.go

```diff
@@ -413,13 +413,18 @@ func (c *KongClient) Update(ctx context.Context) error {
 	// based on the cache contents, we need to ensure it is not modified during the process.
 	var cacheSnapshot store.CacheStores
 	if c.kongConfig.FallbackConfiguration {
-		// TODO: https://github.com/Kong/kubernetes-ingress-controller/issues/6080
-		// Use TakeSnapshotIfChanged to avoid taking a snapshot if the cache hasn't changed.
+		var newSnapshotHash store.SnapshotHash
 		var err error
-		cacheSnapshot, err = c.cache.TakeSnapshot()
+		cacheSnapshot, newSnapshotHash, err = c.cache.TakeSnapshotIfChanged(c.kongConfig.SnapshotHash)
 		if err != nil {
 			return fmt.Errorf("failed to take snapshot of cache: %w", err)
 		}
+		// Empty snapshot hash means that the cache hasn't changed since the last snapshot was taken.
+		if newSnapshotHash == store.SnapshotHashEmpty {
```

Comment:
We could add a note/issue link that will explain that taking a snapshot and optimizing config processing when no changes are detected could be moved out of the `c.kongConfig.FallbackConfiguration` condition and used in a regular code path in the future once we decide it's fine to do so. This could also include a note about the rationale behind keeping it this way for now - something along "we want to isolate new behavior behind a feature gate as much as possible".

Reply:
Note added and dedicated issue created https://github.com/Kong/kubernetes-ingress-controller/issues/6095